### ISSUE TITLE
Investigate Structured Locking

### ISSUE THREAD

**Issue by DanHeidinga** [eclipse/openj9#9314]
From the spec:
> Structured locking is the situation when, during a method invocation, every exit on a given monitor matches a preceding entry on that monitor. Since there is no assurance that all code submitted to the Java Virtual Machine will perform structured locking, implementations of the Java Virtual Machine are permitted but not required to enforce both of the following two rules guaranteeing structured locking. Let T be a thread and M be a monitor. Then:
> 1. The number of monitor entries performed by T on M during a method invocation must equal the number of monitor exits performed by T on M during the method invocation whether the method invocation completes normally or abruptly.
> 2. At no point during a method invocation may the number of monitor exits performed by T on M since the method invocation exceed the number of monitor entries performed by T on M since the method invocation.
> Note that the monitor entry and exit automatically performed by the Java Virtual Machine when  invoking a synchronized method are considered to occur during the calling method's invocation.
-- http://cr.openjdk.java.net/~iris/se/11/latestSpec/java-se-11-jvms-draft-diffs.pdf

So this doesn't look like a structured locking violation to me unless this code is returning a different object:
```
  112 aload0getfield 
  113 getfield 286 org/apache/tools/ant/DirectoryScanner.slowScanLock Ljava/lang/Object;
  116 astore 6
....
  124 aload 6
  126 invokespecial 1387 java/lang/Object.wait()V
```

_Originally posted by @DanHeidinga in https://github.com/eclipse/openj9/issues/9276#issuecomment-617323639_

**Comment by DanHeidinga** [comment]
Issues where structured locking has come up:
* https://github.com/eclipse/openj9/issues/9276
* https://github.com/eclipse/openj9/issues/9267 - would simplify the approach

**Comment by gacholio** [comment]
I suspect to do this efficiently, we should tags methods during verification as not needing checking.

**Comment by DanHeidinga** [comment]
Agreed.  And all return bytecodes - both sync and regular - will need to do the check so there will likely be some startup overhead here

**Comment by gacholio** [comment]
Maybe if we can't prove the structure, we leave returns as generic (and exit the monitor before updating the bytecode). Well-behaved methods won't take any hit then. Might be some shared classes implications for that.

**Comment by DanHeidinga** [comment]
...with a more involved "fixReturns" process for classes in the SharedClasses cache?

**Comment by gacholio** [comment]
Ideally, we would simply stop rewriting the bytecodes in generic return, and rely on it being done by verifier/fixReturns/whatever.

In the structured case, the fact that we successfully exit the monitor on return doesn't imply that every path reaching that return is correct, so fixing it there would be inappropriate anyway.
